compiler/rustc_const_eval/src/const_eval/machine.rs

```diff
@@ -347,6 +347,23 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter<'mir,
                 )?;
                 ecx.write_pointer(ptr, dest)?;
             }
+            sym::const_deallocate => {
+                let ptr = ecx.read_pointer(&args[0])?;
+                let size = ecx.read_scalar(&args[1])?.to_machine_usize(ecx)?;
+                let align = ecx.read_scalar(&args[2])?.to_machine_usize(ecx)?;
+
+                let size = Size::from_bytes(size);
+                let align = match Align::from_bytes(align) {
+                    Ok(a) => a,
+                    Err(err) => throw_ub_format!("align has to be a power of 2, {}", err),
+                };
+
+                ecx.memory.deallocate(
```

Comment:
I think this needs to be slightly more sophisticated -- we need to handle the case where an allocation is created in one `const` and then used (and attempted to be deallocated) in another. Currently this will error, but I think it actually should be a NOP, similar to deallocating a const-allocation at runtime.

Reply:
Catch the deallocation error and ignore it? Or actually check beforehand whether this is a heap allocation?

Reply:
Catching errors is usually a bad idea, so IMO this should be checked beforehand -- if the allocation is in `tcx`, do not call `memory.deallocate`.

Reply:
> Currently this will error, but I think it actually should be a NOP, similar to deallocating a const-allocation at runtime.

So this case would be allowed, is that correct? 

```rust
#![feature(core_intrinsics)]
#![feature(const_heap)]

const X: &u32 = &42;
const Y: () = unsafe {
    let ptr = X as *const u32 as *mut u8;
    core::intrinsics::const_deallocate(ptr, 4, 4);
};

fn main() {
    println!("{} {:?}", X, Y); // 42 ()
}
```

Reply:
Yes. It would be annoying to have to differentiate between previous heap allocations and allocs that never were on the heap. We're allowed to miss some UB if it's too much work to detect it

Reply:
>  It would be annoying to have to differentiate between previous heap allocations and allocs that never were on the heap. 

We could store the `MemoryKind` in the `tcx` allocations... but I agree, we don't have to detect this UB.

Reply:
I've implemented it. (I still think it's a bit counterintuitive...)